canary/generator/generate_anomalies_in_split_data.py

```diff
@@ -0,0 +1,149 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import json
+from collections import defaultdict
+import os
+import pickle
+from copy import deepcopy
+from argparse import ArgumentParser
+
+import numpy as np
+
+
+def aggregate(data, buckets, field):
+    """
+    Aggregates the data by the field
+    :param data: split data for one metric
+    :param buckets: buckets for one metric
+    :param field: field by which to aggregate
+    :return: aggregated data
+    """
+    agg_data = defaultdict(
+        lambda: defaultdict(lambda: defaultdict(lambda: np.zeros(len(buckets))))
+    )
+    for opts in data:
+        for date, hist in opts['data'].items():
+            try:
+                agg_data[opts[field]]['data'][date] += hist
+            except:
+                import ipdb; ipdb.set_trace()
+    for arch, data in agg_data.items():
+        agg_data[arch]['buckets'] = buckets
+        agg_data[arch]['data'] = {date: hist/np.sum(hist)
+                                  for date, hist in data['data'].items()}
+    return agg_data
+
+
+def defaultdict_to_dict(def_dict):
+    """
+    Changes nested defaultdict into nested dict. It's necessary for example for pickling.
+    :param def_dict: defaultdict to be changed
+    :return: nested dict with the same structure as input defaultdict
+    """
+    if isinstance(def_dict, (defaultdict, dict)):
+        normal_dict = dict(def_dict)
+        for k, v in normal_dict.items():
+            normal_dict[k] = defaultdict_to_dict(v)
+        return normal_dict
+    else:
+        return def_dict
+
+
+if __name__ == '__main__':
+    parser = ArgumentParser()
+    parser.add_argument(dest='data_dir',
+                        help='Directory with previously generated data', nargs='*')
+    parser.add_argument('-b', '--split_data_dir', dest='split_data_dir',
+                        help='Directory with downloaded split data', nargs='*')
+    parser.add_argument('-s', '--save_dir', dest='save_dir',
+                        help='Directory where the transformed data should be saved')
+    args = parser.parse_args()
+    X_test = dict()
```

Comment:
nit: any reason not to use {} instead of dict()? No need to change it, just curious.

Reply:
pycharm thinks it's better :P